File: 2021/go/day02.go
package main

import (
    "os"
    "bufio"
    "fmt"
    "strconv"
    "strings"
)

type command struct {
    direction string
    dist int
}

type coord struct {
    x int
    y int
}

func get_input() []command {
    commands := make([]command, 0)

    scanner := bufio.NewScanner(os.Stdin)

    for {
        scanner.Scan()
        raw_command := scanner.Text()

        if len(raw_command) == 0 { break }

        split_command := strings.Split(raw_command, " ")
        dist, _ := strconv.Atoi(split_command[1])

        commands = append(commands, command{
            direction: split_command[0],
            dist: dist,
        })
    }

    return commands
}

func solve_a(commands []command) int {
    pos := coord{ x: 0, y: 0}

    for _, command := range commands {
        switch command.direction {
            case "forward":
                pos.x += command.dist
            case "down":
                pos.y += command.dist
            case "up":
                pos.y -= command.dist
        }
    }

    return pos.x * pos.y
}

func solve_b(commands []command) int {
    pos := coord{ x: 0, y: 0}
    aim := 0

    for _, command := range commands {
        switch command.direction {
            case "forward":
                pos.x += command.dist
                pos.y += command.dist * aim
            case "down":
                aim += command.dist
            case "up":
                aim -= command.dist
        }
    }

    return pos.x * pos.y
}


func main() {
    numbers := get_input()

    fmt.Println("Solution A-part", solve_a(numbers))
    fmt.Println("Solution B-part", solve_b(numbers))
}

File: 2021/rs/src/bin/day02.rs
mod day02 {
    fn execute_course_a(course: &[(String, i32)]) -> (i32, i32) {
        let mut pos = (0, 0);
        for (command, x) in course {
            pos = match command.as_str() {
                "forward" => { (pos.0 + x, pos.1) },
                "down" => { (pos.0, pos.1 + x) },
                "up" => { (pos.0, pos.1 - x) },
                _ => unreachable!(),
            };
        }

        pos
    }

    fn execute_course_b(course: &[(String, i32)]) -> (i32, i32) {
        let mut aim = 0;
        let mut pos = (0, 0);
        for (command, x) in course {
            match command.as_str() {
                "forward" => { pos = (pos.0 + x, pos.1 + x * aim) },
                "down" => { aim += x },
                "up" => { aim -= x },
                _ => unreachable!(),
            };
        }

        pos
    }

    pub fn solve_a(course: &[(String, i32)]) -> i32 {
        let final_pos = execute_course_a(course);
        final_pos.0 * final_pos.1
    }

    pub fn solve_b(course: &[(String, i32)]) -> i32 {
        let final_pos = execute_course_b(course);
        final_pos.0 * final_pos.1
    }
}

fn main() {
    use std::io::{self, BufRead};

    let stdin = io::stdin();

    let course: Vec<(String, i32)> = stdin.lock().lines()
        .map(|s| {
            let command: Vec<String> = s.unwrap().split(' ').map(|s| s.into()).collect();
            (String::from(&command[0]), command[1].parse().unwrap())})
        .collect();

    println!("Solution A-part: {}", day02::solve_a(&course));
    println!("Solution B-part: {}", day02::solve_b(&course));
}
